### PR TITLE
Feature/hcap 1287 - Set/Edit allocations

### PR DESCRIPTION
**Pull request by mmumby**
# work completed:
## DB
- dropped `start_date` and `end_date` from `site_phase_allocation`

## FE
- added form to set/edit Allocations
- Populated start/end date on form with phase dates - not editable.
- Added Validation (Must be a positive number, required)

## BE
- added POST and PATCH
- added BE validation 
- returned `allocation_id` to the phase payload to allow editing on FE
- removed references to allocation `start_date` and `end_date`

## Testing
- Cypress tests for happy path, and UI validations
- Server and E2E tests testing happy path, unauthorized, data validation for POST and PATCH

## UI
Screen Shot 2023-02-14 at 12 32 16 PM
Screen Shot 2023-02-14 at 12 32 56 PM

### CODE REVIEW

File: server/routes/allocation.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+  ],
+  asyncMiddleware(async (req, resp) => {
+    if (!FEATURE_PHASE_ALLOCATION) {
+      return resp.status(501).send('Phase allocation feature not active');
+    }
+    try {
+      const { body, hcapUserInfo: user } = req;
+      const response = await updatePhaseAllocation(req.params.id, body, user);
+      logger.info({
+        action: 'phase-allocation_post',
```
[the remark this answers is not in the document]
Comment (mmumby) [reply server/routes/allocation.js:69]: I was following the pattern for creating global phases (which I guess isn't fully implemented yet) but it was using a patch. I can make it a put!

File: server/services/allocations.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+ * @property {number} site_id       FK for employer_site
+ */
+
+const getPhaseAllocation = async (siteId, phaseId) => {
+  const allocation = await dbClient.db[collections.SITE_PHASE_ALLOCATION].findOne(
+    {
+      site_id: siteId,
+      phase_id: phaseId,
+    },
+    { limit: 1 }
```
[the remark this answers is not in the document]
Comment (mmumby) [reply server/services/allocations.js:20]: no - forgot to remove it after changing it from `find` to `findOne`, I'll update it!

File: server/routes/allocation.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+  })
+);
+
+// Update phase allocation: PATCH
+router.patch(
+  '/:id',
+  [
+    keycloak.allowRolesMiddleware('ministry_of_health'),
+    keycloak.getUserInfoMiddleware(),
+    expressRequestBodyValidator(CreateAllocationSchema),
```
[the remark this answers is not in the document]
Comment (mmumby) [reply server/routes/allocation.js:59]: phase_id/site_id are included in the BE validations (not the FE though) 

```
phase_id: yup.number().required('Allocation must be related to a phase'),
site_id: yup.number().required('Allocation must be related to a site'),
```

File: server/routes/allocation.js [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+          username: user.username,
+          id: user.id,
+        },
+        allocationId: response.id,
+      });
+      logger.info(response);
+      return resp.status(201).json(response);
+    } catch (err) {
+      logger.error(err);
+      return resp.status(400).send(`${err}`);
```
[the remark this answers is not in the document]
Comment (mmumby) [reply server/routes/allocation.js]: Updated to be a specific message,  but appears to show "Bad Request" in the toast message regardless of what is sent. I will have to investigate further.

File: server/routes/allocation.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (mmumby) [reply server/routes/allocation.js:69]: Just realized this is not what you were referring to... updated the sting to match the request... but should it be a PATCH or PUT?

[the remark this answers is not in the document]
Comment (fw-noel) [reply server/routes/allocation.js]: Patch is good! I just meant the logger info :)

File: server/routes/allocation.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (fw-noel) [reply server/routes/allocation.js]: Yeah it's less what it shows right now but more what it could potentially show to the user, whether its in an error message or just the network tab. If there's a possibility that it returns something like
"Error on UPDATE table_name SET" etc. etc., not wonderful.

Updating to a specific message resolves this comment :)

File: server/routes/allocation.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (mmumby) [reply server/routes/allocation.js]: switched it back to PATCH!

File: server/routes/allocation.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (fw-noel) [reply server/routes/allocation.js:59]: Sorry yes, I was wondering if using the same Schema for the EDIT endpoint works okay since phase_id and site_id is expected, but I don't think we want to / need to be passing those in?

[the remark this answers is not in the document]
Comment (mmumby) [reply server/routes/allocation.js:59]: ohh good point, I was passing those in but it's not necessary. I'll update the patch/validation for that

[the remark this answers is not in the document]
Comment (mmumby) [reply server/routes/allocation.js]: updated!

File: server/migrations/1676328102966_drop-allocation-dates.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
+/* eslint-disable camelcase */
+const { collections } = require('../db');
+
+exports.shorthands = 'drop-allocation-dates';
+
+exports.up = (pgm) => {
+  pgm.dropColumns(collections.SITE_PHASE_ALLOCATION, ['start_date', 'end_date'], {
```
Comment (rhattie-ey) [inline server/migrations/1676328102966_drop-allocation-dates.js:7]: I appreciate this migration, makes things tidier.

File: server/services/allocations.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
+/* eslint-disable camelcase */
+// disabling camelcase check so that we can manipulate snake_case attributes without changing structure
+const { dbClient, collections } = require('../db');
+
+/**
+ *
```
Comment (rhattie-ey) [inline server/services/allocations.js:6]: I like that there's a type being defined, but I don't see any references being made to it yet. Is that going to be done in another commit? I'm also wondering if there's a clearer, more unique type name than `allocation` to give it, but that's a nitpick.

File: server/routes/allocation.js [hunk 7]
```diff
@@ -0,0 +1,85 @@
+const express = require('express');
```
Comment (rhattie-ey) [inline server/routes/allocation.js:1]: Having files called both `allocation` and `phase-allocation` is confusing, as are the endpoints of the same name. I'd propose coming up with a better distinction between the two - perhaps `phase-allocation` is the actual allocation stuff (as you are doing here), and the global stuff becomes `global-phases` or something? I'm open to ideas on how to rename these.

File: client/src/services/allocations.js [hunk 8]
```diff
@@ -0,0 +1,28 @@
+import store from 'store';
+import { API_URL } from '../constants';
+
+export const createPhaseAllocation = async (allocationJson) => {
```
Comment (rhattie-ey) [inline client/src/services/allocations.js:4]: I know it's a convention established in other files, but the fact that a parameter called `allocationJson` isn't actually JSON bothers me. I'd suggest renaming it to something like `allocationData`, to better hint at its type.

**Review by rhattie-ey — COMMENTED** [review]
Looks good so far! No major complaints from me.

File: server/tests/util/keycloak.js [hunk 9]
```diff
@@ -21,6 +22,11 @@ const healthAuthority = {
   password: process.env.KC_TEST_HA_PWD || 'password',
 };
 
+const ministryOfHealth = {
+  username: 'test-moh',
+  password: 'password',
```
[the remark this answers is not in the document]
Comment (rhattie-ey) [reply server/tests/util/keycloak.js]: Resolved in latest batch of commits.

File: server/services/phase.js [hunk 10]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable */
```
[the remark this answers is not in the document]
Comment (mmumby) [reply server/services/phase.js:1]: I meant to remove this -  I think I wanted to push to GitHub to test a theory/bypass linting and forgot to remove it. Thanks for catching it, removed it now!

File: server/routes/allocation.js (hunk 7 above)
Comment (mmumby) [reply server/routes/allocation.js:1]: This is exactly what I wanted feedback on - it was getting confusing for me too. I originally started naming everything 'PhaseAllocations' to match the db table/FK relationship (without using SitePhaseAllocation because that seemed too long). But eventually, I simplified everything to just 'allocation' as there appeared to be some overlap with global phases (Which has a few different names in the codebase -ie  phase/phases, sitePhases, globalPhases)

I can rename a few things - I kind of think allocations and phases can stand alone without any extra context in their names. 
ie - createAllocation, updateAllocation, getPhase, createPhase, updatePhase. Thoughts?

File: client/src/services/allocations.js (hunk 8 above)
Comment (mmumby) [reply client/src/services/allocations.js]: renamed to payload!

File: server/services/allocations.js (hunk 6 above)
Comment (mmumby) [reply server/services/allocations.js:6]: I think perhaps I am confused, I was using this as a doc block to define what is being returned from `getPhaseAllocation`, What am I missing?

File: server/services/phase.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (rhattie-ey) [reply server/services/phase.js]: Just a heads up, you also left the same directives in `allocations.service.test` and `allocation.e2e.test`, as of the current commit.

File: server/routes/allocation.js (hunk 7 above)
Comment (fw-noel) [reply server/routes/allocation.js:1]: I think `phases` vs. `allocations` would be the way to go. I wouldn't think global phase, because there's a chance it won't be fully "global" later on.

File: server/services/allocations.js (hunk 6 above)
Comment (rhattie-ey) [reply server/services/allocations.js:6]: `@typedef` is a JSDoc tag which *defines* a type. It does not mark any given symbols as that type. For that, you need `@type`. Your JSDoc block is equivalent to the following TypeScript:

```typescript
type allocation {
  id: number;
  phase_id: string;
  allocation: number;
  site_id: number;
}
```

In other words, you're making a new *type* at the file scope, which happens to share a name with a local *variable*, but has nothing to do with it. We can check this by hovering over the `allocation` variable:
image

This can be fixed by tagging the variable with `@type {typename}`, like so:
image

And now, VSCode and other IDEs can provide information as you edit, and TS can check typing properly:
image
image

File: client/src/services/allocations.js (hunk 8 above)
Comment (rhattie-ey) [reply client/src/services/allocations.js]: Much better!

File: server/routes/allocation.js (hunk 7 above)
Comment (mmumby) [reply server/routes/allocation.js:1]: If this is the agreed path forward, I can rename all the allocation stuff in this PR, and create a separate PR/ticket to rename phases.

File: server/services/allocations.js (hunk 6 above)
Comment (mmumby) [reply server/services/allocations.js]: ah - okay thank you for the detailed explanation, I think I have it working the correct way now.

File: server/services/phase.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (mmumby) [reply server/services/phase.js]: removed!

File: server/routes/allocation.js (hunk 7 above)
Comment (rhattie-ey) [reply server/routes/allocation.js:1]: That works for me!

File: server/services/allocations.js [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+ */
+const getAllocation = async (siteId, phaseId) => {
+  /**
+   * @typedef {Object} allocationResponse  type for DB response to the `getAllocation` query
+   * @property {number} id                 PK ID of the allocation
+   * @property {number} phase_id           FK for phase
+   * @property {number} allocation         Number of allocations available
+   * @property {number} site_id            FK for employer_site
+   *
+   *
```
Comment (rhattie-ey) [inline server/services/allocations.js:18]: Nitpick: these extra lines can be removed.

**Review by rhattie-ey — DISMISSED** [review]
Looks good to me!

File: server/validators/allocation.js [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+const { validateBlankOrPositiveInteger } = require('./helpers');
+
+const CreateAllocationSchema = yup
+  .object()
+  .noUnknown('Unknown field in entry')
+  .shape({
+    allocation: yup
+      .number()
+      .nullable()
+      .max(99, 'Must be between 0-99')
```
Comment (fw-noel) [inline server/validators/allocation.js:11]: Ah shoot sorry I meant to say, no max limits!
Please remove the related test too!

Comment (mmumby) [reply server/validators/allocation.js]: removed!

File: server/services/allocations.js (hunk 11 above)
Comment (mmumby) [reply server/services/allocations.js]: removed!

**Review by fw-noel — APPROVED** [review]

**Review by rhattie-ey — APPROVED** [review]
Looks good, approved!

**Review by vanics — APPROVED** [review]
